postgres-protocol/src/password/mod.rs

```diff
@@ -0,0 +1,102 @@
+//! Functions to encrypt a password in the client.
+//!
+//! This is intended to be used by client applications that wish to
+//! send commands like `ALTER USER joe PASSWORD 'pwd'`. The password
+//! need not be sent in cleartext if it is encrypted on the client
+//! side. This is good because it ensures the cleartext password won't
+//! end up in logs pg_stat displays, etc.
+
+use crate::authentication::sasl;
+use hmac::{Hmac, Mac, NewMac};
+use md5::Md5;
+use rand::RngCore;
+use sha2::digest::FixedOutput;
+use sha2::{Digest, Sha256};
+
+#[cfg(test)]
+mod test;
+
+const SCRAM_DEFAULT_ITERATIONS: u32 = 4096;
+const SCRAM_DEFAULT_SALT_LEN: usize = 16;
+
+/// Error encrypting password.
+#[derive(Debug)]
+pub struct EncryptionError(());
+
+/// Encrypt password using SCRAM-SHA-256 with a randomly-generated
+/// salt.
+///
+/// An error will be raised if the password contains characters
+/// prohibited by [RFC 4013](https://tools.ietf.org/html/rfc4013).
+///
+/// The client may assume the returned string doesn't contain any
+/// special characters that would require escaping in an SQL command.
+pub fn scram_sha_256(password: &str) -> Result<String, EncryptionError> {
+    let mut salt: [u8; SCRAM_DEFAULT_SALT_LEN] = [0; SCRAM_DEFAULT_SALT_LEN];
+    let mut rng = rand::thread_rng();
+    rng.fill_bytes(&mut salt);
+    scram_sha_256_salt(password, salt)
+}
+
+/// Encrypt password using SCRAM-SHA-256 with a caller-provided
+/// salt.
+///
+/// This is useful if the caller has access to a better source
+/// of randomness.
+///
+/// An error will be raised if the password contains characters
+/// prohibited by [RFC 4013](https://tools.ietf.org/html/rfc4013).
```

Comment:
For backwards compatibility, Postgres's version of SCRAM works with arbitrary byte sequences and just skips the saslprep step for invalid ones: https://github.com/sfackler/rust-postgres/blob/master/postgres-protocol/src/authentication/sasl.rs#L20-L23.

So this function should probably just take a `&[u8]` instead of a `&str` to support that.

Reply:
Done. This answers some of the other problems, as well, because now scram_sha_256 can't return an error. I still left the functions separate, because it seems like slightly better ergonomics.

Reply:
I think you may need to push again?

Reply:
Pushed, sorry.